cudax/include/cuda/experimental/__memory_resource/any_resource.cuh

```diff
@@ -42,17 +42,150 @@
 {
 template <class _Ty, class _Uy = _CUDA_VSTD::remove_cvref_t<_Ty>>
 _LIBCUDACXX_INLINE_VAR constexpr bool __is_basic_any_resource = false;
 
+template <class _Resource,
+          _CUDA_VMR::_AllocType =
+            _CUDA_VMR::async_resource<_Resource> ? _CUDA_VMR::_AllocType::_Async : _CUDA_VMR::_AllocType::_Default>
+struct _Shared_resource
+{
+  static_assert(_CUDA_VMR::resource<_Resource>, "");
+
+  template <class... _Args>
+  explicit _Shared_resource(_Args&&... __args)
+      : __control_block(new _Control_block{_Resource{_CUDA_VSTD::forward<_Args>(__args)...}, 1})
+  {}
+
+  _Shared_resource(const _Shared_resource& __other) noexcept
+      : __control_block(__other.__control_block)
+  {
+    __control_block->__ref_count.fetch_add(1, _CUDA_VSTD::memory_order_relaxed);
+  }
+
+  _Shared_resource(_Shared_resource&& __other) noexcept
+      : __control_block(_CUDA_VSTD::exchange(__other.__control_block, nullptr))
+  {}
+
+  ~_Shared_resource()
+  {
+    if (__control_block && __control_block->__ref_count.fetch_sub(1, _CUDA_VSTD::memory_order_acq_rel) == 1)
+    {
+      delete __control_block;
+    }
+  }
+
+  _Shared_resource& operator=(const _Shared_resource& __other) noexcept
+  {
+    if (this != &__other)
+    {
+      _Shared_resource(__other).swap(*this);
+    }
+
+    return *this;
+  }
+
+  _Shared_resource& operator=(_Shared_resource&& __other) noexcept
+  {
+    if (this != &__other)
+    {
+      _Shared_resource(_CUDA_VSTD::move(__other)).swap(*this);
+    }
+
+    return *this;
+  }
+
+  void swap(_Shared_resource& __other) noexcept
+  {
+    _CUDA_VSTD::swap(__control_block, __other.__control_block);
+  }
+
+  _CCCL_NODISCARD void* allocate(size_t __size, size_t __alignment)
+  {
+    return __control_block->__resource.allocate(__size, __alignment);
+  }
+
+  void deallocate(void* __ptr, size_t __size, size_t __alignment)
+  {
+    __control_block->__resource.deallocate(__ptr, __size, __alignment);
+  }
+
+  friend void swap(_Shared_resource& __left, _Shared_resource& __right) noexcept
+  {
+    __left.swap(__right);
+  }
+
+  _CCCL_NODISCARD_FRIEND bool operator==(const _Shared_resource& __left, const _Shared_resource& __right)
+  {
+    if (__left.__control_block == __right.__control_block)
+    {
+      return true;
+    }
+
+    if (__left.__control_block == nullptr || __right.__control_block == nullptr)
+    {
+      return false;
+    }
+
+    return __left.__control_block->__resource == __right.__control_block->__resource;
+  }
+
+  _CCCL_NODISCARD_FRIEND bool operator!=(const _Shared_resource& __left, const _Shared_resource& __right)
+  {
+    return !(__left == __right);
+  }
+
+  _LIBCUDACXX_TEMPLATE(class _Property)
+  _LIBCUDACXX_REQUIRES(::cuda::has_property<_Resource, _Property>)
+  _CCCL_NODISCARD_FRIEND __property_value_t<_Property> get_property(const _Shared_resource& __self, _Property) noexcept
+  {
+    return get_property(__self.__control_block->__resource, _Property{});
+  }
```

Comment:
Important: This is missing the valueless property forward. The reason being that `__property_value_t` does not work on a property that has no `value_type`, like `cuda::mr::device_accessible`